### PR TITLE
[Model] Add GIN Model

### PR DESCRIPTION
**Pull request by kitaev-chen**
# Description
This PR implement dgl version of GIN model according to the paper of "How Powerful are Graph Neural Networks" and its original code, and together with an on-the-fly dataset and adapted data loader for the GIN dataset which is available in https://github.com/weihua916/powerful-gnns/blob/master/dataset.zip.

# Checklist
Please feel free to remove inapplicable items for your PR.

- [x]  The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x]    Changes are complete (i.e. I finished coding on this PR)
- [x]    All changes have test coverage
- [x]    Code is well-documented
- [x]    To the my best knowledge, examples are either not affected by this change,
or have been fixed to be compatible with this change
- [x]    Related issue is referred in this PR

# Changes
Essentially, this PR uses DGL's API to implement GIN model. The gin layer is mainly handled by DGL, plus the re-weight procedure to update node's feature.

### CODE REVIEW

**Comment by jermainewang** [comment]
Hey @kitaev-chen , cool work and thanks for the contributing! You need to remove the data files in the PR and download them on-the-fly (see instruction [here](https://docs.dgl.ai/contribute.html#data-hosting)). BTW, what kind of dataset is that?

**Comment by kitaev-chen** [comment]
> Hey @kitaev-chen , cool work and thanks for the contributing! You need to remove the data files in the PR and download them on-the-fly (see instruction [here](https://docs.dgl.ai/contribute.html#data-hosting)). BTW, what kind of dataset is that?

OK I will see the instruction of data part. 

This dataset includes some popular graph kernel data such as collab, imdb, mutag, nci1, proteins, ptc and reddit which listed on [graph kernel datasets](https://ls11-www.cs.tu-dortmund.de/staff/morris/graphkerneldatasets), but obviously the author has already processed them and make them more compact. 

For example, in the smallest dataset MUTAG, the first five line are like this:

188    # 188 graphs
23 2    # the first graph includes 23 nodes and the graph label is 2
2 2 1 13    # the first node of first graph has a node label 2 and 2 edges which are node 1 and 13
2 2 0 2    
2 3 1 3 11    # the third node of first graph has a node label 2 and 3 edges which are node 1, 3 and 11

Hope it is easy to adapt to the data api of dgl.

**Comment by VoVAllen** [comment]
I've made one PR for TUDataset, which is at https://github.com/dmlc/dgl/pull/473.
Could you have a look and see whether that's fine in your case?

**Comment by kitaev-chen** [comment]
> I've made one PR for TUDataset, which is at #473.
> Could you have a look and see whether that's fine in your case?

Thanks for you work! It will be very helpful for researchers. 

But for GIN case, the dataset is more compact into 1 .txt or .mat file.  Considering there will be many works which want to do benchmark with that best ICLR paper, it's better to have a new GINDataset. I have already convert dataset.py to it.

**Comment by jermainewang** [comment]
I think it's fair to have dataset for quickly reproduce and compare results of the GIN paper. If user wants more control, the TUDataset can come in handy. I see value in both. @VoVAllen could you please review this PR?

File: python/dgl/data/gindt.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+_url = 'https://raw.githubusercontent.com/weihua916/powerful-gnns/master/dataset.zip'
+
+
+class GINDataset(object):
+    """Datasets for Graph Isomorphism Network (GIN)
+    Adapted for the paper 'how powerful are graph neural networks?'
+
+    The dataset contains the compact format of popular graph kernel datasets,
+    which included:
+    MUTAG, COLLAB, IMDBBINARY, IMDBMULTI, NCI1, PROTEINS, PTC, REDDITBINARY, REDDITMULTI5K
```
Comment (jermainewang) [inline python/dgl/data/gindt.py]: Maybe add a pointer to `TUDataset` so user will know how to access other graph datasets?

Comment (kitaev-chen) [reply python/dgl/data/gindt.py]: > Maybe add a pointer to `TUDataset` so user will know how to access other graph datasets?

good idea

Comment (kitaev-chen) [reply python/dgl/data/gindt.py]: there is an error when checked cpp code but I never touch cpp part, see detail https://discuss.dgl.ai/t/category-build-errors-found-when-checking-by-task-lint/186

Comment (kitaev-chen) [reply python/dgl/data/gindt.py]: > Maybe add a pointer to `TUDataset` so user will know how to access other graph datasets?
"
This datset class processes all data sets listed above. For more graph kernel datasets, 
    see :class:`TUDataset`  
"
done

**Comment by VoVAllen** [comment]
Hi, we expect our dataset to be backend agnostic for both pytorch and mxnet user. Therefore could you change the tensor to numpy matrix and convert them to tensor in your collate function? Thanks!

**Comment by kitaev-chen** [comment]
> Hi, we expect our dataset to be backend agnostic for both pytorch and mxnet user. Therefore could you change the tensor in the GINDataset to numpy matrix and convert them to tensor in your collate function? Thanks!

Sure, I will work on that.

**Comment by VoVAllen** [comment]
Hi, 

Could you change the `logging.info` to `tqdm.write`? Currently the print information is a bit messy due to the combination of logging and tqdm.

Also could you add your self to our [contributor](https://github.com/dmlc/dgl/blob/master/CONTRIBUTORS.md) list? 

By the way, I'm wondering whether the result is comparable to the original paper?

Others looks good to me. Thanks for your contribution!

**Comment by kitaev-chen** [comment]
> Hi,
> 
> Could you change the `logging.info` to `tqdm.write`? Currently the print information is a bit messy due to the combination of logging and tqdm.
> 
> Also could you add your self to our [contributor](https://github.com/dmlc/dgl/blob/master/CONTRIBUTORS.md) list?
> 
> By the way, I'm wondering whether the result is comparable to the original paper?
> 
> Others looks good to me. Thanks for your contribution!

Done. You're welcome.

**Comment by VoVAllen** [comment]
Thanks for your contribution!
